Objects/unicodeobject.c

```diff
@@ -14617,6 +14617,62 @@ unicode_new_impl(PyTypeObject *type, PyObject *x, const char *encoding,
     return unicode;
 }
 
+static const char *
+arg_as_utf8(PyObject *obj, const char *name)
+{
+    if (!PyUnicode_Check(obj)) {
+        PyErr_Format(PyExc_TypeError,
+                     "str() argument '%s' must be str, not %T",
+                     name, obj);
+        return NULL;
+    }
+    const char *str = _PyUnicode_AsUTF8NoNUL(obj);
+    if (str == NULL) {
+        return NULL;
+    }
+    return str;
+}
+
+static PyObject *
+unicode_vectorcall(PyObject *type, PyObject *const *args,
+                   size_t nargsf, PyObject *kwnames)
+{
+    assert(Py_Is(_PyType_CAST(type), &PyUnicode_Type));
+
+    Py_ssize_t nargs = PyVectorcall_NARGS(nargsf);
+    if (kwnames != NULL && PyTuple_GET_SIZE(kwnames) != 0) {
+        // Fallback to unicode_new()
+        PyObject *tuple = _PyTuple_FromArray(args, nargs);
+        if (tuple == NULL) {
+            return NULL;
+        }
+        PyObject *dict = _PyStack_AsDict(args + nargs, kwnames);
+        if (dict == NULL) {
+            Py_DECREF(tuple);
+            return NULL;
+        }
+        PyObject *ret = unicode_new(_PyType_CAST(type), tuple, dict);
+        Py_DECREF(tuple);
+        Py_DECREF(dict);
+        return ret;
+    }
+    if (nargs > 3) {
+        PyErr_Format(PyExc_TypeError,
+            "str() takes at most 3 arguments (%d given)", nargs);
+        return NULL;
+    }
```

Comment:
@corona10: this would be the actual change.

```suggestion
    if (!_PyArg_CheckPositional("str", nargs, 0, 3)) {
        return NULL;
    }
```

Reply:
Yeah, yeah, I was lazy :)

Reply:
Note that this would change the exception message. Currently we have this:

```console
$ python3
Python 3.13.0a6 (v3.13.0a6:57aee2a02c, Apr  9 2024, 09:56:43) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> str(1, 1, 1, 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
    str(1, 1, 1, 1)
    ~~~^^^^^^^^^^^^
TypeError: str() takes at most 3 arguments (4 given)
```

With `_PyArg_CheckPositional` we get this:
```console
$ ./python.exe
Python 3.13.0a6+ (heads/perf/unicode-vectorcall-pos-only-dirty:8138db40dd, Apr 11 2024, 15:03:45) [Clang 15.0.0 (clang-1500.3.9.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> str(1, 1, 1, 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
    str(1, 1, 1, 1)
    ~~~^^^^^^^^^^^^
TypeError: str expected at most 3 arguments, got 4
```

Perhaps it does not matter.

Reply:
FYI, All vectorcall implementation is using _PyArg_CheckPositional as possible.

Reply:
@corona10, not `enumerate` ;)

Reply:
> @corona10, not enumerate ;)

Yeah, there are some exceptional case, but please don't generalize it :)

> Perhaps it does not matter.

Yeah, We change the same things for other objects.

Reply:
oh, also not `long`.

Reply:
... nor `range` :)

Reply:
But sure, yes, the majority of them use `_PyArg_CheckPositional`.